homeassistant/config_entries.py

```diff
@@ -1599,25 +1617,34 @@ async def _async_migrate_func(
     ) -> dict[str, Any]:
         """Migrate to the new version."""
         data = old_data
-        if old_major_version == 1 and old_minor_version < 2:
-            # Version 1.2 implements migration and freezes the available keys
-            for entry in data["entries"]:
-                # Populate keys which were introduced before version 1.2
-
-                pref_disable_new_entities = entry.get("pref_disable_new_entities")
-                if pref_disable_new_entities is None and "system_options" in entry:
-                    pref_disable_new_entities = entry.get("system_options", {}).get(
-                        "disable_new_entities"
+        if old_major_version == 1:
+            if old_minor_version < 2:
+                # Version 1.2 implements migration and freezes the available keys
+                for entry in data["entries"]:
+                    # Populate keys which were introduced before version 1.2
+
+                    pref_disable_new_entities = entry.get("pref_disable_new_entities")
+                    if pref_disable_new_entities is None and "system_options" in entry:
+                        pref_disable_new_entities = entry.get("system_options", {}).get(
+                            "disable_new_entities"
+                        )
+
+                    entry.setdefault("disabled_by", entry.get("disabled_by"))
+                    entry.setdefault("minor_version", entry.get("minor_version", 1))
+                    entry.setdefault("options", entry.get("options", {}))
+                    entry.setdefault(
+                        "pref_disable_new_entities", pref_disable_new_entities
+                    )
+                    entry.setdefault(
+                        "pref_disable_polling", entry.get("pref_disable_polling")
                     )
+                    entry.setdefault("unique_id", entry.get("unique_id"))
```

Comment:
- structural pattern matching to simplify the value extraction logic
- `setdefault` already doesn't override the key if it exists in a `dict`, so no need to re-extract pre-existing values. Additionally, the default value to populate it with is already `None`, so that can be omitted.

```suggestion
                    match entry:
                        case {"pref_disable_new_entities": None, "system_options": {"disable_new_entities": disable_new_entities}}:
                            pref_disable_new_entities = disable_new_entities
                        case {"pref_disable_new_entities": old_pref_disable_new_entities}:
                            pref_disable_new_entities = old_pref_disable_new_entities

                    entry.setdefault("disabled_by")
                    entry.setdefault("minor_version", 1)
                    entry.setdefault("options", {})
                    entry.setdefault("pref_disable_new_entities", pref_disable_new_entities)
                    entry.setdefault("pref_disable_polling")
                    entry.setdefault("unique_id")
```

Reply:
I changed nothing on 1.2 migration except splitting out `if old_major_version == 1:`.
Refactoring of the 1.2 can be done if needed in a follow up PR